openmc/mgxs/mdgxs.py

```diff
@@ -0,0 +1,2619 @@
+            The type of multi-delayed-group cross section object to return
+        domain : openmc.Material or openmc.Cell or openmc.Universe or
+            openmc.RegularMesh
+            The domain for spatial homogenization
+        domain_type : {'material', 'cell', 'distribcell', 'universe', 'mesh'}
+            The domain type for spatial homogenization
+        energy_groups : openmc.mgxs.EnergyGroups
+            The energy group structure for energy condensation
+        by_nuclide : bool
+            If true, computes cross sections for each nuclide in domain.
+            Defaults to False
+        name : str, optional
+            Name of the multi-group cross section. Used as a label to identify
+            tallies in OpenMC 'tallies.xml' file. Defaults to the empty string.
+        delayed_groups : list of int
+            Delayed groups to filter out the xs
+        num_polar : Integral, optional
+            Number of equi-width polar angle bins for angle discretization;
+            defaults to one bin
+        num_azimuthal : Integral, optional
+            Number of equi-width azimuthal angle bins for angle discretization;
+            defaults to one bin
+        Returns
+        -------
+        openmc.mgxs.MDGXS
+            A subclass of the abstract MDGXS class for the multi-delayed-group
+            cross section type requested by the user
+
+        """
+
+        cv.check_value('mdgxs_type', mdgxs_type, MDGXS_TYPES)
+
+        if mdgxs_type == 'delayed-nu-fission':
+            mdgxs = DelayedNuFissionXS(domain, domain_type, energy_groups,
+                                       delayed_groups)
+        elif mdgxs_type == 'chi-delayed':
+            mdgxs = ChiDelayed(domain, domain_type, energy_groups,
+                               delayed_groups)
+        elif mdgxs_type == 'beta':
+            mdgxs = Beta(domain, domain_type, energy_groups, delayed_groups)
+        elif mdgxs_type == 'decay-rate':
+            mdgxs = DecayRate(domain, domain_type, energy_groups, delayed_groups)
+        elif mdgxs_type == 'delayed-nu-fission matrix':
+            mdgxs = DelayedNuFissionMatrixXS(domain, domain_type, energy_groups,
+                                             delayed_groups)
+
+        mdgxs.by_nuclide = by_nuclide
+        mdgxs.name = name
+        mdgxs.num_polar = num_polar
+        mdgxs.num_azimuthal = num_azimuthal
+        return mdgxs
+
+    def get_xs(self, groups='all', subdomains='all', nuclides='all',
+               xs_type='macro', order_groups='increasing',
+               value='mean', delayed_groups='all', squeeze=True, **kwargs):
+        """Returns an array of multi-delayed-group cross sections.
+
+        This method constructs a 4D NumPy array for the requested
+        multi-delayed-group cross section data for one or more
+        subdomains (1st dimension), delayed groups (2nd demension),
```

Comment:
Thank you for clearly laying out the dimensionality of the arrays.  Super helpful for review!